pkg/http/listener.go

```diff
@@ -0,0 +1,326 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package http
+
+import (
+	"crypto/tls"
+	"errors"
+	"net"
+	"net/http"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+)
+
+var sslRequiredErrMsg = []byte("HTTP/1.0 403 Forbidden\r\n\r\nSSL required")
+
+// HTTP methods.
+var methods = []string{
+	http.MethodGet,
+	http.MethodHead,
+	http.MethodPost,
+	http.MethodPut,
+	http.MethodPatch,
+	http.MethodDelete,
+	http.MethodConnect,
+	http.MethodOptions,
+	http.MethodTrace,
+	"PRI", // HTTP 2 method
+}
+
+// maximum length of above methods + one space.
+var methodMaxLen = getMethodMaxLen() + 1
+
+func getMethodMaxLen() int {
+	maxLen := 0
+	for _, method := range methods {
+		if len(method) > maxLen {
+			maxLen = len(method)
+		}
+	}
+
+	return maxLen
+}
+
+func isHTTPMethod(s string) bool {
+	for _, method := range methods {
+		if s == method {
+			return true
+		}
+	}
+
+	return false
+}
+
+type acceptResult struct {
+	conn net.Conn
+	err  error
+}
+
+// httpListener - HTTP listener capable of handling multiple server addresses.
+type httpListener struct {
+	mutex                  sync.Mutex         // to guard Close() method.
+	tcpListeners           []*net.TCPListener // underlaying TCP listeners.
+	acceptCh               chan acceptResult  // channel where all TCP listeners write accepted connection.
+	doneChs                []chan struct{}    // done channels for each goroutine for TCP listeners.
```

Comment:
Each tcpListener handler goroutine is controlled by a separate doneCh. The individual doneCh is closed on `Close()`.  If there is one doneCh, multiple goroutines read in parallel. I am not sure reading from one channel by multiple goroutine is safe.